### ISSUE TITLE
Defining `define_editor` overrides default editor support

### ISSUE THREAD

**Issue by jakobnissen** [JuliaLang/julia#37231]
To reproduce:
```
julia> InteractiveUtils.define_editor("vi", wait=true) do cmd, path, line
           `$cmd +$line $path`
       end
 1-element Array{Function,1}:
 #2 (generic function with 1 method)

julia> ENV["JULIA_EDITOR"] = "nano" # changed my mind, I want to use nano instead
"nano"

julia> @edit 1+1
ERROR: no editor found
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] edit(::String, ::Int32) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/InteractiveUtils/src/editless.jl:205
 [3] edit(::Function, ::Any) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.5/InteractiveUtils/src/editless.jl:223
 [4] top-level scope at REPL[34]:1
```
What happens is that the default editors are only defined on the first call of `edit`, iff there are no defined editors. I can't see the rationale for that. Wouldn't it be nice to keep default editors in any case?

**Comment by apparluk** [comment]
if using ed start as

```
shell> ed filename
```

vi and vim don't accept ~/ in the path

vi

```
bad line number                                                       Command
```

vim
```
"~/.julia/config/startup.jl" [NEW DIRECTORY]                   0,01       All
```

```
bash-4.3$ echo $JULIA_EDITOR

bash-4.3$ julia --banner=no
julia> InteractiveUtils.editor()
`emacs`

julia> ENV["EDITOR"] = "nano"
"nano"

julia> edit("~/.julia/config/startup.jl", 2)

julia> ENV["EDITOR"] = "vi"
"vi"

julia> edit("/home/eric/.julia/config/startup.jl", 2)
reading /home/eric/.julia/config/startup.jl












julia> ENV["EDITOR"] = "vim"
"vim"

julia> edit("/home/eric/.julia/config/startup.jl", 2)

julia> exit()
bash-4.3$ 
```

**Comment by StefanKarpinski** [comment]
> What happens is that the default editors are only defined on the first call of `edit`, iff there are no defined editors. I can't see the rationale for that. Wouldn't it be nice to keep default editors in any case?

I don't think that behavior is intentional, I think it was just intended to run `define_default_editors` once and only once. The fact that it has the effect of preventing `define_default_editors` from being called if someone has already defined a different editor seems accidental.

**Comment by StefanKarpinski** [comment]
@apparluk, I'm not seeing how this comment is related?

**Comment by jakobnissen** [comment]
Can we just define the default editors in global scope so it happens when InteractiveUtils is loaded? Or, if there are problem with that I can't forsee, we could have some global variable `DEFAULT_EDITORS_DEFINED = false`, and check that?

**Comment by apparluk** [comment]
@StefanKarpinski

> vi and vim don't accept ~/ in the path

That might be a simple fix in the Julia code.

The manual might mention that ed and nvi are unsupported.

**Comment by StefanKarpinski** [comment]
Those are totally unrelated issues. Please open a separate issue if you want to discuss those.

**Comment by StefanKarpinski** [comment]
Not sure what we could possibly do about elvis lacking wide character support?

**Comment by apparluk** [comment]
@StefanKarpinski

I won't open an issue on vim accepting ~/ in the path started from UXTerm but not the REPL.

> elvis lacking wide character support

Looking at their website it appears for three years there has been no active development.

**Comment by StefanKarpinski** [comment]
In any case, this issue is not the place to talk about these unrelated things.
